README.md
# route53-dyndns-service

CHANGES.rst
Changelog (route53-dyndns-service)
==================================

0.0.1 (Unreleased)
--------------------

- Initial project skeleton


tests/helpers.py
def new_resource_record(name, value):
    return {
        'Name': name,
        'Type': 'A',
        'SetIdentifier': 'string',
        'Region': 'us-east-1',
        'TTL': 123,
        'ResourceRecords': [
            {
                'Value': value
            },
        ],
    }


tests/test_backend.py
from datetime import datetime
import unittest

from route53_dyndns.route53 import (
    find_resource_record, Route53Exception, update_resource_record)

from mock import patch

from .helpers import new_resource_record


class MockRoute53Client(object):
    def _resource_record(self, name, value):
        return {
            'Name': name,
            'Type': 'A',
            'SetIdentifier': 'string',
            'Region': 'us-east-1',
            'TTL': 123,
            'ResourceRecords': [
                {
                    'Value': value
                },
            ],
        }

    def _resource_records_response(self, name, max_items):
        return {
            'ResourceRecordSets': [
                new_resource_record(name, 'string')
            ],
            'IsTruncated': False,
            'MaxItems': max_items
        }

    def list_resource_record_sets(self, HostedZoneId=None,
                                  StartRecordName=None, MaxItems=None):
        assert HostedZoneId
        assert StartRecordName
        assert MaxItems

        return self._resource_records_response(StartRecordName, MaxItems)

    def change_resource_record_sets(self, HostedZoneId=None,
                                    ChangeBatch=None):
        assert HostedZoneId
        assert ChangeBatch

        return {
            'ChangeInfo': {
                'Id': 'string',
                'Status': 'PENDING',
                'SubmittedAt': datetime(2015, 1, 1),
                'Comment': ChangeBatch['Comment']
            }
        }


class BackendTestCase(unittest.TestCase):
    def test_find_resource_record(self):
        """ Test find_resource_record functionality """

        client = MockRoute53Client()
        hostname = "www.google.com"

        # Test some invalid inputs
        with self.assertRaises(ValueError):
            find_resource_record(None, client=client)

        with self.assertRaises(ValueError):
            find_resource_record('', client=client)

        # Go right case
        resource_record = find_resource_record(hostname, client=client)
        self.assertEqual(resource_record['Name'], hostname)

        # Test a go wrong case when there's a Route 53 error
        with patch.object(client, 'list_resource_record_sets') as mocked:
            mocked.side_effect = RuntimeError("Route 53 Error")

            with self.assertRaises(Route53Exception):
                find_resource_record(hostname, client=client)

        # Test a go wrong case where the returned result is truncated
        truncated_result = client._resource_records_response(hostname, 1)
        truncated_result['IsTruncated'] = True

        with patch.object(client, 'list_resource_record_sets') as mocked:
            mocked.return_value = truncated_result

            with self.assertRaises(AssertionError):
                find_resource_record(hostname, client=client)

        # Test a go wrong case where there are multiple resource record results
        multi_result = client._resource_records_response(hostname, 1)
        multi_result['ResourceRecordSets'] += new_resource_record(hostname,
                                                                  'whoops')

        with patch.object(client, 'list_resource_record_sets') as mocked:
            mocked.return_value = multi_result

            with self.assertRaises(AssertionError):
                find_resource_record(hostname, client=client)

    def test_update_resource_record(self):
        """ Test update_resource_record functionality """

        client = MockRoute53Client()
        hostname = "www.google.com"
        value = '127.0.0.1'
        record = new_resource_record(hostname, value)

        # Go right case where the value is the same as the in the record
        updated = update_resource_record(record, value, client=client)
        self.assertTrue(updated)

        # Go right case
        updated = update_resource_record(record, '192.168.1.1', client=client)
        self.assertTrue(updated)

        # Test a go wrong case when there's a Route 53 error
        with patch.object(client, 'change_resource_record_sets') as mocked:
            mocked.side_effect = RuntimeError("Route 53 Error")

            with self.assertRaises(Route53Exception):
                update_resource_record(record, '192.168.1.1', client=client)


tests/test_frontend.py
from __future__ import unicode_literals

from base64 import b64encode
import unittest

from flask import Response

from route53_dyndns import app, route53, views

from mock import patch

from .helpers import new_resource_record


class FrontendTestCase(unittest.TestCase):
    def setUp(self):
        self.app = app.test_client()
        self.url = '/nic/update'

    def make_auth_header(self, username, password):
        value = b64encode("{0}:{1}".format(username, password).encode('ascii'))

        return {
            'Authorization': b'Basic ' + value
        }

    def get_with_auth(self, url, username='admin', password='admin', **kwargs):
        auth_header = self.make_auth_header(username, password)

        return self.app.get(url, headers=auth_header, **kwargs)

    def assertResponseEqual(self, expected, rv):
        response = rv.get_data().decode('utf-8')

        return self.assertEqual(expected + '\r\n', response)

    def test_app_response(self):
        response = "Hello World!"

        # Check that a string response always has correct type and newline
        rv = app.make_response(response)
        self.assertEqual('text/plain', rv.headers['content-type'])
        self.assertResponseEqual(response, rv)

        # Check that a tuple response always has correct type and newline
        rv = app.make_response((response, 200, None))
        self.assertEqual('text/plain', rv.headers['content-type'])
        self.assertResponseEqual(response, rv)

        # Check that a response gets the correct type
        rv = app.make_response(Response(response, 200, None))
        self.assertEqual('text/plain', rv.headers['content-type'])

    def test_verify_auth(self):
        verified = views.verify_auth('admin', 'admin')
        self.assertFalse(verified)

        verified = views.verify_auth('admin', 'secret')
        self.assertTrue(verified)

    def test_nic_update_http(self):
        # Check that the trailing slash on the URL is optional
        rv = self.app.get(self.url + '/?hostname=foo.com')
        self.assertNotEqual(rv.status_code, 404)

        rv = self.app.get(self.url + '?hostname=foo.com')
        self.assertNotEqual(rv.status_code, 404)

        # Test the authentication logic for the view
        with patch('route53_dyndns.views.verify_auth') as mocked:
            auth_url = self.url + '?hostname=foo.com'

            # Test no authentication
            rv = self.app.get(auth_url)
            self.assertEqual(rv.status_code, 401)

            # Test incorrect authentication
            mocked.return_value = False
            rv = self.get_with_auth(auth_url)
            self.assertEqual(rv.status_code, 403)

            # Test correct authentication
            mocked.return_value = True
            rv = self.get_with_auth(auth_url)
            self.assertEqual(rv.status_code, 200)

    @patch('route53_dyndns.views.verify_auth', **{'method.return_value': True})
    def test_nic_update_parameters(self, mocked_auth):
        # Offline is an unsupported parameter
        rv = self.get_with_auth(self.url + '?offline=True')
        self.assertEqual(rv.status_code, 200)
        self.assertResponseEqual(views.NOT_SUPPORTED, rv)

    @patch('route53_dyndns.views.verify_auth', **{'method.return_value': True})
    def test_nic_update_hostname(self, mocked_auth):
        # Test no hostname
        rv = self.get_with_auth(self.url)
        self.assertEqual(rv.status_code, 200)
        self.assertResponseEqual(views.NO_HOST, rv)

        # Test a go wrong case when there's a Route 53 error
        with patch('route53_dyndns.route53.find_resource_record') as mocked:
            mocked.side_effect = route53.Route53Exception("Error")

            rv = self.get_with_auth(self.url + '?hostname=foo')
            self.assertResponseEqual(views.GENERAL_ERROR, rv)

        # Test hostname not found
        with patch('route53_dyndns.route53.find_resource_record') as mocked:
            mocked.return_value = None

            rv = self.get_with_auth(self.url + '?hostname=foo')
            self.assertResponseEqual(views.NO_HOST, rv)

    @patch('route53_dyndns.route53.find_resource_record')
    @patch('route53_dyndns.views.verify_auth', **{'method.return_value': True})
    def test_nic_update_record_exists(self, mocked_auth, mocked_find_record):
        hostname = "www.google.com"
        value = "10.1.10.1"
        new_value = "192.168.1.1"
        mocked_find_record.return_value = new_resource_record(hostname, value)

        # Test IP hasn't changed
        url = self.url + '?hostname=' + hostname + '&myip=' + value
        rv = self.get_with_auth(url)
        self.assertResponseEqual(views.NO_CHANGE % value, rv)

        # Test IP hasn't changed, implicit IP from request
        rv = self.get_with_auth(self.url + '?hostname=' + hostname,
                                environ_base={'REMOTE_ADDR': value})
        self.assertResponseEqual(views.NO_CHANGE % value, rv)

        # Test go wrong case when there's an exception
        with patch('route53_dyndns.route53.update_resource_record') as mocked:
            mocked.side_effect = route53.Route53Exception("Error")

            url = self.url + '?hostname=' + hostname + '&myip=' + new_value
            rv = self.get_with_auth(url)
            self.assertResponseEqual(views.GENERAL_ERROR, rv)

        # Test go wrong case when there's an expected failure
        with patch('route53_dyndns.route53.update_resource_record') as mocked:
            mocked.return_value = False

            url = self.url + '?hostname=' + hostname + '&myip=' + new_value
            rv = self.get_with_auth(url)
            self.assertResponseEqual(views.GENERAL_ERROR, rv)

        with patch('route53_dyndns.route53.update_resource_record') as mocked:
            mocked.return_value = True

            # Go right case with explicit new IP
            url = self.url + '?hostname=' + hostname + '&myip=' + new_value
            rv = self.get_with_auth(url)
            self.assertResponseEqual(views.IP_CHANGED % new_value, rv)

            # Go right case with implict new IP from request
            url = self.url + '?hostname=' + hostname + '&myip=' + new_value
            rv = self.get_with_auth(url,
                                    environ_base={'REMOTE_ADDR': new_value})
            self.assertResponseEqual(views.IP_CHANGED % new_value, rv)


route53_dyndns/cmdline.py
#!/usr/bin/env python

""" Command-line functionality for route53_dyndns """

import logging
import sys

from optparse import OptionParser

from route53_dyndns import __version__, app

logger = logging.getLogger(__name__)


def configure_logging(level=logging.INFO, logger=None):
    """ Configure logging for the command-line """

    if logger is None:  # pragma: no cover
        logger = logging.getLogger()

    handler = logging.StreamHandler(sys.stdout)

    logger.addHandler(handler)
    logger.setLevel(level)


def main(args=sys.argv[1:], parser=None):
    """ Main entry point for the route53_dyndns command-line script """

    usage = "%prog [options] [args]\n\n"
    version = "%prog " + __version__

    usage += "%prog is the HTTP dynamic DNS server for Route53"

    if parser is None:  # pragma: no cover
        parser = OptionParser(usage=usage, version=version,
                              add_help_option=False)
    parser.remove_option("--version")  # Remove standard version option

    parser.add_option("-h", "--help", dest="help", action="store_true",
                      help="Show help and exit")
    parser.add_option("-V", "--version", dest="version", action="store_true",
                      help="Show version and exit")

    (options, parsed_args) = parser.parse_args(args)

    # Check if we should print help or version number
    if not parsed_args or (parsed_args and len(parsed_args) == 0):
        if options.version:
            parser.print_version()
            sys.exit(0)
        elif options.help:
            parser.print_help()
            sys.exit(0)

    configure_logging()

    app.run(debug=True)


route53_dyndns/route53.py
""" Backend functionality for communicating with Route 53 and updating DNS """

import boto3


class Route53Exception(Exception):
    pass


def get_client():  # pragma: no cover
    """ Helper function to get an authenticated Route 53 client """

    return boto3.client('route53')


def find_resource_record(record_name, client=None):
    """ Find a DNS resource record on Route 53 """

    if not client:  # pragma: no cover
        client = get_client()

    if not record_name:
        raise ValueError("Need a record name")

    try:
        response = client.list_resource_record_sets(
            HostedZoneId='string',
            StartRecordName=record_name,
            MaxItems='1'
        )
    except Exception as e:
        # TODO - Log the error
        raise Route53Exception(e)

    # We don't ever expect to get truncated results since we are looking up a
    # single record by name. TBD: This may fail if record types can share names
    assert not response['IsTruncated'], "Expected single record"
    assert len(response['ResourceRecordSets']) == 1, "Expected single record"

    return response['ResourceRecordSets'][0]


def update_resource_record(resource_record, value, client=None):
    """ Update a resource record on Route 53 with a new value """

    if not client:  # pragma: no cover
        client = get_client()

    record_name = resource_record['Name']
    record_type = resource_record['Type']
    set_identifier = resource_record['SetIdentifier']
    current_value = resource_record['ResourceRecords'][0]['Value']

    if value == current_value:
        return True  # Nothing to do, it is already up-to-date

    try:
        client.change_resource_record_sets(
            HostedZoneId='string',
            ChangeBatch={
                'Comment': "Updating DNS record via route53_dyndns",
                'Changes': [
                    {
                        'Action': 'UPSERT',
                        'ResourceRecordSet': {
                            'Name': record_name,
                            'Type': record_type,
                            'SetIdentifier': set_identifier,
                            'ResourceRecords': [
                                {
                                    'Value': value
                                },
                            ],
                        }
                    },
                ]
            }
        )
    except Exception as e:
        # TODO - Log the error
        raise Route53Exception(e)

    return True


route53_dyndns/views.py
""" HTTP API for updating dynamic DNS records """

from __future__ import unicode_literals

from functools import wraps

from flask import request

from route53_dyndns.app import app
from route53_dyndns import route53

AUTH_REALM = "Route 53 DNS Update API"

BAD_AUTH = 'badauth'
IP_CHANGED = 'good %s'
NO_CHANGE = 'nochg %s'
NO_HOST = 'nohost'
NOT_SUPPORTED = '!donator'
GENERAL_ERROR = '911'


def verify_auth(username, password):
    """ Verify the HTTP Basic Auth credentials """

    # TODO - Replace with auth from config
    return username == 'admin' and password == 'secret'


def authenticate_response(forbidden=False):
    """ Generate an authenticate response for HTTP Basic Auth """

    headers = {
        'WWW-Authenticate': 'Basic realm="{}"'.format(AUTH_REALM)
    }

    return (BAD_AUTH, 403 if forbidden else 401, headers)


def api_auth(view):
    """ Decorator to add HTTP Basic Auth to a view """

    @wraps(view)
    def decorated(*args, **kwargs):
        auth = request.authorization

        if not auth:  # Auth is required at all times
            return authenticate_response()

        if not verify_auth(auth.username, auth.password):
            return authenticate_response(forbidden=True)

        return view(*args, **kwargs)

    return decorated


@app.route('/nic/update/', methods=['GET'])
@app.route('/nic/update', methods=['GET'])
@api_auth
def nic_update():
    """ Update a dynamic DNS record """

    if request.args.get('offline'):
        return NOT_SUPPORTED

    # TODO - Add user agent check

    try:
        hostname = request.args.get('hostname')
        resource_record = route53.find_resource_record(hostname)
    except ValueError:
        return NO_HOST
    except:
        return GENERAL_ERROR

    if not resource_record:
        return NO_HOST

    # TODO - A sanity check of the value would be good, but good luck doing it
    #        easily in anything before Python 3.3+ which got 'ipaddress' module
    # TODO - Comma separated values should be allowed
    myip = request.args.get('myip', request.remote_addr)

    if myip == resource_record['ResourceRecords'][0]['Value']:
        return NO_CHANGE % myip

    try:
        if not route53.update_resource_record(resource_record, myip):
            return GENERAL_ERROR
    except:
        return GENERAL_ERROR

    return IP_CHANGED % myip
